### PR TITLE
st1 [1782][ADD] stock_carrier_info

### PR DESCRIPTION
**Pull request by TomohiroAbe-Quartile**
[1782](https://www.quartile.co/web#id=1782&action=771&model=project.task&view_type=form&menu_id=505)

### CODE REVIEW

**Review by yostashiro — COMMENTED** [review]
@TomohiroAbe-Quartile Can we change the module name to stock_carrier?  Please change the name of the added model to stock.carrier, as well.

**Comment by yostashiro** [comment]
> Can we change the module name to stock_carrier? Please change the name of the added model to stock.carrier, as well.

Sorry on second thought, 'stock_carrier_info'/'stock.carrier.info', because of the concern on the field name clash with the existing `carrier_id`.  We can add `carrier_info_id` to stock.picking.

File: delivery_master_mi7/readme/DESCRIPTION.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+This module does the following:
+
+- Add delivery_master model.
+- Add fields in stock_picking/carrier_data view.
```
Comment (yostashiro) [inline delivery_master_mi7/readme/DESCRIPTION.rst]:
```suggestion
- Adds a new model stock.carrier.info to keep the carrier name and the tracking URL against the carrier code.
- Adds carrier_info_id field in stock.picking.

The module does not contain any business logic per se.
```

Comment (TomohiroAbe-Quartile) [reply delivery_master_mi7/readme/DESCRIPTION.rst]: I edited. Thank you.

**Comment by TomohiroAbe-Quartile** [comment]
@yostashiro 
The following has been achieved.
Functions:
1. Send Carrier Info and URL by E-mail
2. Add Stock Carrier Info menu in delivery.menu_delivery
3. Compute carrier name and url in account_invoice

But now,  stock_delivery_result_import doesn't worked.
Sorry, I want to your help how to import stock_carrier_info.

File: stock_carrier_info/models/account_invoice.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# -*- coding: utf-8 -*-
```
Comment (yostashiro) [inline stock_carrier_info/models/account_invoice.py]: The content of this file should be moved to account_invoice_validate_send_email.  Pay attention to the comment "The module does not contain any business logic per se."

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: stock_carrier_info/models/account_invoice.py (hunk 2 above)
Comment (TomohiroAbe-Quartile) [reply stock_carrier_info/models/account_invoice.py]: Thank you, I will be careful.
Moved logics to account_invoice_validate_send_email and it worked as expect.

**Comment by TomohiroAbe-Quartile** [comment]
Thank you, I read a code and understand how to work.

File: account_invoice_validate_send_email/models/account_invoice.py [hunk 3]
```diff
@@ -23,6 +23,17 @@ class AccountInvoice(models.Model):
         help="Delivery slip numbers taken from the linked deliveries.",
         compute="_compute_carrier_tracking_refs",
     )
+    carrier_info_name = fields.Char(
+        "Delivery Carrier Name",
+        help="Delivery Carrier Information for send a e-mail to customer.",
+        compute="_compute_carrier_info",
+    )
```
Comment (yostashiro) [inline account_invoice_validate_send_email/models/account_invoice.py]:
```suggestion
    carrier_name = fields.Char(
        compute="_compute_carrier_info",
    )
```

File: account_invoice_validate_send_email/models/account_invoice.py [hunk 4]
```diff
@@ -23,6 +23,17 @@ class AccountInvoice(models.Model):
...
+
+    carrier_info_url = fields.Char(
+        "Delivery Carrier URL",
+        help="Delivery Carrier's URL for tracking.",
+        compute="_compute_carrier_info",
+    )
```
Comment (yostashiro) [inline account_invoice_validate_send_email/models/account_invoice.py]:
```suggestion
    carrier_tracking_url = fields.Char(
        compute="_compute_carrier_info",
    )
```

File: account_invoice_validate_send_email/models/account_invoice.py [hunk 5]
```diff
@@ -126,3 +137,11 @@ def _compute_carrier_tracking_refs(self):
...
                 if x.carrier_tracking_ref
             )
+
+    @api.multi
+    def _compute_carrier_info(self):
+        for invoice in self:
+            picking = invoice.picking_ids[0]
+            if picking.carrier_info_id:
+                invoice.carrier_info_name = picking.carrier_info_id.name
+                invoice.carrier_info_url = picking.carrier_info_id.tracking_url
```
Comment (yostashiro) [inline account_invoice_validate_send_email/models/account_invoice.py]:
```suggestion
        for invoice in self:
            carrier_recs = invoice.picking_ids.mapped("carrier_info_id")
            tracking_refs = invoice.picking_ids.mapped("carrier_tracking_ref")
            invoice.carrier_name = ", ".join(x.name for x in carrier_recs)
            invoice.carrier_tracking_url = ", ".join(x.tracking_url for x in carrier_recs)
            invoice.carrier_tracking_refs = ", ".join(tracking_refs)
```
Does this make sense (note that I haven't tested the code)? This should replace `_compute_carrier_tracking_refs()` as well.

Currently `_compute_carrier_tracking_refs()` takes care of the cases where an invoice involves multiple pickings. If that is the design we are after, we should also apply the same idea for the carrier name and the URL fields to take care of multiple pickings.

File: account_invoice_validate_send_email/views/account_invoice_view.xml [hunk 6]
```diff
@@ -9,6 +9,10 @@
                 <field name="invoice_sent" />
                 <field name="web_url" invisible="1" />
             </xpath>
+            <xpath expr="//field[@name='move_id']" position="after">
+                <field name="carrier_info_name" />
+                <field name="carrier_info_url" />
+            </xpath>
```
Comment (yostashiro) [inline account_invoice_validate_send_email/views/account_invoice_view.xml]: I suppose it's not necessary to show the carrier fields. We currently do not show the tracking refs either.

File: stock_carrier_info/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    "name": "Stock Carrier Info",
+    "version": "10.0.1.0.0",
+    "category": "Stock",
+    "license": "AGPL-3",
+    "author": "Quartile Limited",
+    "website": "https://www.quartile.co/",
+    "depends": ["sale_stock"],
+    "data": [
+        "security/ir.model.access.csv",
+        "security/carrier_info_group.xml",
```
Comment (yostashiro) [inline stock_carrier_info/__manifest__.py]:
```suggestion
```

File: stock_carrier_info/security/carrier_info_group.xml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo noupdate="1">
+    <record id="group_carrier_info" model="res.groups">
+        <field name="name">Stock Carrier Info</field>
+        <field name="category_id" ref="base.module_category_hidden" />
+    </record>
+</odoo>
```
Comment (yostashiro) [inline stock_carrier_info/security/carrier_info_group.xml]: Remove the file.

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: account_invoice_validate_send_email/models/account_invoice.py (hunk 5 above)
Comment (TomohiroAbe-Quartile) [reply account_invoice_validate_send_email/models/account_invoice.py]: Thank you, I added code and confirmed work.

File: stock_carrier_info/security/carrier_info_group.xml (hunk 8 above)
Comment (TomohiroAbe-Quartile) [reply stock_carrier_info/security/carrier_info_group.xml]: Thank you. I removed this file.

File: account_invoice_validate_send_email/views/account_invoice_view.xml (hunk 6 above)
Comment (TomohiroAbe-Quartile) [reply account_invoice_validate_send_email/views/account_invoice_view.xml]: Sorry, I just added this line for debug.
I delited it.

File: account_invoice_validate_send_email/models/account_invoice.py [hunk 9]
```diff
@@ -116,13 +119,13 @@ def action_send(self):
                 self.invoice_sent = True
         return True
 
-    # Return cases, where there is no direct link from the invoice to the pickings,
-    # are outside of the scope of showing the tracking references to the customer.
```
Comment (yostashiro) [inline account_invoice_validate_send_email/models/account_invoice.py]: This comment should still be valid.

File: account_invoice_validate_send_email/models/account_invoice.py [hunk 10]
```diff
@@ -21,8 +21,11 @@ class AccountInvoice(models.Model):
     carrier_tracking_refs = fields.Char(
         "Tracking References",
         help="Delivery slip numbers taken from the linked deliveries.",
-        compute="_compute_carrier_tracking_refs",
+        compute="_compute_carrier_info",
     )
+    carrier_info_name = fields.Char(compute="_compute_carrier_info",)
+
```
Comment (yostashiro) [inline account_invoice_validate_send_email/models/account_invoice.py]:
```suggestion
```
I prefer to get rid of blank lines unless there is a good reason not to do so.

File: account_invoice_validate_send_email/views/account_invoice_view.xml [hunk 11]
```diff
@@ -7,7 +7,6 @@
         <field name="arch" type="xml">
             <xpath expr="//field[@name='move_id']" position="after">
                 <field name="invoice_sent" />
-                <field name="web_url" invisible="1" />
```
Comment (yostashiro) [inline account_invoice_validate_send_email/views/account_invoice_view.xml]: No need to delete this one, I suppose, although it's not a big deal.

**Review by yostashiro — APPROVED** [review]
LGTM with some non-blocking comments. 👍

**Pull request merged**